src/net_processing.cpp

```diff
@@ -2392,6 +2392,20 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
                 }
                 pindexWalk = pindexWalk->pprev;
             }
+            // Special case for second cmpctblock request of tip
+            // Must match all conditions to add to vToFetch above (except mmapBlocksInFlight.count)
+            // and conditions to do CMPCT_BLOCK announcement below.
+            if (vToFetch.size() == 0 && pindexLast->pprev == chainActive.Tip() &&
+                mmapBlocksInFlight.size() == mmapBlocksInFlight.count(pindexLast->GetBlockHash()) &&
+                mmapBlocksInFlight.count(pindexLast->GetBlockHash()) < MAX_CMPCTBLOCKS_INFLIGHT_PER_BLOCK &&
+                !(pindexLast->nStatus & BLOCK_HAVE_DATA) &&
+                (!IsWitnessEnabled(pindexLast->pprev, chainparams.GetConsensus()) || State(pfrom->GetId())->fHaveWitness) &&
+                nodestate->fSupportsDesiredCmpctVersion) {
+                vToFetch.push_back(pindexLast);
```

Comment:
I believe the above needs fixing in three ways - MarkBlockAsInFlight needs to be more robust against NULL pit, the request needs to move into this if statement and skip the remainder of this block of code, and we shouldn't double-request from the same peer.